apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,381 @@
+graph generated by `aot_autograd` for the above program.
+
+```python
+def forward(self, arg0_1: f32[3, 3]):
+    # File: test/test_scratch.py:334, code: return torch.sin(x) + torch.square(x)
+    sin: f32[3, 3] = torch.ops.aten.sin.default(arg0_1)
+    pow_1: f32[3, 3] = torch.ops.aten.pow.Tensor_Scalar(arg0_1, 2);  arg0_1 = None
+    add: f32[3, 3] = torch.ops.aten.add.Tensor(sin, pow_1);  sin = pow_1 = None
+    return (add,)
+
+```
+
+inductor: As discussed above, it is inductor's job to apply optimisations and
+generate specialised code. In this case, it has fused `sin` and `square` to run
+within the same `for`-loop. This allows the generated program to do more compute
+per read/write effectively improving the memory bandwith utilization.
+
+```python
+from ctypes import c_void_p, c_long
+import torch
+import math
+import random
+import os
+import tempfile
+from math import inf, nan
+from torch._inductor.hooks import run_intermediate_hooks
+from torch._inductor.utils import maybe_profile
+
+from torch import empty_strided, device
+from torch._inductor.codecache import AsyncCompile
+from torch._inductor.select_algorithm import extern_kernels
+
+aten = torch.ops.aten
+assert_size_stride = torch._C._dynamo.guards.assert_size_stride
+reinterpret_tensor = torch.ops.inductor._reinterpret_tensor
+async_compile = AsyncCompile()
+
+
+cpp_fused_add_cos_sin_0 = async_compile.cpp('''
+#include "/tmp/torchinductor_kshiteej/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
+extern "C" void kernel(const float* in_ptr0,
+                       float* out_ptr0)
+{
+    {
+        for(long i0=static_cast<long>(0L); i0<static_cast<long>(8L); i0+=static_cast<long>(8L))
+        {
+            auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<long>(i0));
+            auto tmp1 = tmp0.sin();
+            auto tmp2 = tmp0.cos();
+            auto tmp3 = tmp1 + tmp2;
+            tmp3.store(out_ptr0 + static_cast<long>(i0));
+        }
+        #pragma omp simd simdlen(4) 
+        for(long i0=static_cast<long>(8L); i0<static_cast<long>(9L); i0+=static_cast<long>(1L))
+        {
+            auto tmp0 = in_ptr0[static_cast<long>(i0)];
+            auto tmp1 = std::sin(tmp0);
+            auto tmp2 = std::cos(tmp0);
+            auto tmp3 = tmp1 + tmp2;
+            out_ptr0[static_cast<long>(i0)] = tmp3;
+        }
+    }
+}
+''')
+
+
+async_compile.wait(globals())
+del async_compile
+
+def call(args):
+    arg0_1, = args
+    args.clear()
+    assert_size_stride(arg0_1, (3, 3), (3, 1))
+    buf0 = empty_strided((3, 3), (3, 1), device='cpu', dtype=torch.float32)
+    cpp_fused_add_cos_sin_0(c_void_p(arg0_1.data_ptr()), c_void_p(buf0.data_ptr()))
+    del arg0_1
+    return (buf0, )
+```
+
+### Teaching `dynamo` about `torch.func` transforms
+
+Now that we have a basic understanding of how `torch.compile` works, let's
+delve into how we can extend the support for `torch.func` transforms. Given that
```

Comment:
Is this something the user is expected to do, or is it something it's implemented in `torch.compile` and the user can now simply use?